pkg/util/kms/kms_kmip_storage.go

```diff
@@ -0,0 +1,479 @@
+		conn.Close()
+		return nil, err
+	}
+
+	// KMIP handshake
+	if err = k.discover(conn); err != nil {
+		conn.Close()
+		return nil, err
+	}
+	return conn, nil
+}
+
+// Send KMIP operation over tls connection, return response and error
+func (k *KMIPSecretStorage) send(conn *tls.Conn, operation kmip14.Operation, payload interface{}) (*kmip.ResponseMessage, *ttlv.Decoder, []byte, error) {
+	biID := uuid.New()
+
+	msg := kmip.RequestMessage{
+		RequestHeader: kmip.RequestHeader{
+			ProtocolVersion: kmip.ProtocolVersion{
+				ProtocolVersionMajor: protocolMajor,
+				ProtocolVersionMinor: protocolMinor,
+			},
+			BatchCount: 1,
+		},
+		BatchItem: []kmip.RequestBatchItem{
+			{
+				UniqueBatchItemID: biID[:],
+				Operation:         operation,
+				RequestPayload:    payload,
+			},
+		},
+	}
+
+	req, err := ttlv.Marshal(msg)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	_, err = conn.Write(req)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	decoder := ttlv.NewDecoder(bufio.NewReader(conn))
+	resp, err := decoder.NextTTLV()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var respMsg kmip.ResponseMessage
+	err = decoder.DecodeValue(&respMsg, resp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return &respMsg, decoder, biID[:], nil
+}
+
+// Verify the response success and return the batch item
+func (k *KMIPSecretStorage) response(respMsg *kmip.ResponseMessage, operation kmip14.Operation, uniqueBatchItemID []byte) (*kmip.ResponseBatchItem, error) {
+	if respMsg.ResponseHeader.BatchCount != 1 {
+		return nil, fmt.Errorf("Batch count %v should be 1", respMsg.ResponseHeader.BatchCount)
+	}
+	if len(respMsg.BatchItem) != 1 {
+		return nil, fmt.Errorf("Batch Intems list len %v should be 1", len(respMsg.BatchItem))
+	}
+	bi := respMsg.BatchItem[0]
+	if operation != bi.Operation {
+		return nil, fmt.Errorf("Unexpected operation, real %v expected %v", bi.Operation, operation)
+	}
+	if !bytes.Equal(uniqueBatchItemID, bi.UniqueBatchItemID) {
+		return nil, fmt.Errorf("Unexpected uniqueBatchItemID, real %v expected %v", bi.UniqueBatchItemID, uniqueBatchItemID)
+	}
+	if kmip14.ResultStatusSuccess != bi.ResultStatus {
+		return nil, fmt.Errorf("Unexpected result status %v expected success %v", bi.ResultStatus, kmip14.ResultStatusSuccess)
+	}
+
+	return &bi, nil
+}
+
+// KMIP handshake
+func (k *KMIPSecretStorage) discover(conn *tls.Conn) error {
+	respMsg, decoder, uniqueBatchItemID, err := k.send(conn, kmip14.OperationDiscoverVersions, kmip.DiscoverVersionsRequestPayload{
+		ProtocolVersion: []kmip.ProtocolVersion{
+			{ProtocolVersionMajor: protocolMajor, ProtocolVersionMinor: protocolMinor},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	bi, err := k.response(respMsg, kmip14.OperationDiscoverVersions, uniqueBatchItemID)
+	if err != nil {
+		return err
+	}
+
+	var respDiscoverVersionsPayload kmip.DiscoverVersionsResponsePayload
+	err = decoder.DecodeValue(&respDiscoverVersionsPayload, bi.ResponsePayload.(ttlv.TTLV))
```

Comment:
```suggestion
	ttlvPayload, ok := bi.ResponsePayload.(ttlv.TTLV)
	if !ok {
		return "", errors.New("failed to parse responsePayload")
	}

	err = decoder.DecodeValue(&respDiscoverVersionsPayload, ttlvPayload)
```

add a typecheck here ?
 
(a linter was complaining over at cephcsi)